Homework5/Routers/Routers/RoutersNetwork.cs

```diff
@@ -0,0 +1,133 @@
+                {
+                    var value = int.Parse(s[1..s.LastIndexOf(')')]);
+                    _edges.Add(new Edge(vertex0, vertex1, weight: value));
+                }
+            }
+        }
+    }
+    
+    /// <summary>
+    /// Writes optimal routers network in file
+    /// </summary>
+    public int BuildNetwork(string targetPath)
+    {
+        var processedVertices = new bool[_verticesCount + 1]; 
+        var outputLines = new StringBuilder[_verticesCount];
+        FillOutputPattern(outputLines);
+        UpdateAvailableEdges(1, processedVertices);
+        processedVertices[1] = true;
+        for (var i = 0;  i < _verticesCount - 1; i++)
+        {
+            var newTreeEdge = _availableEdges.Extract();
+            //  из _availableEdges нельзя просто так взять максимальное ребро, нужно еще раз проверить,
+            // что 1 вершина помечена, а другая нет
+            while (processedVertices[newTreeEdge.Vertex0] && processedVertices[newTreeEdge.Vertex1])
+            {
+                newTreeEdge = _availableEdges.Extract();
+            }
+            
+            // если не нашлось подходящего ребра но мы все ещё в цикле, то граф был несвязный
+            if (newTreeEdge.Weight == 0)
+            {
+                var ew = Console.Error;
+                ew.Write("routers network were unconnected");
+                return -1;
+            }
+            
+            if (processedVertices[newTreeEdge.Vertex1]) // если вершина 1 помечена 
+            {
+                processedVertices[newTreeEdge.Vertex0] = true;
+                UpdateAvailableEdges(newTreeEdge.Vertex0, processedVertices);
+            }
+            else // если вершина 0 помечена
+            {
+                processedVertices[newTreeEdge.Vertex1] = true;
+                UpdateAvailableEdges(newTreeEdge.Vertex1, processedVertices);
+            }
+            
+            TreeEdges.Add(newTreeEdge, newTreeEdge.Weight);
+            outputLines[newTreeEdge.Vertex0 - 1].Append($" {newTreeEdge.Vertex1} ({newTreeEdge.Weight}),");
+        }
+        
+        Print(targetPath, outputLines);
+        return 0;
+    }
+
+    private void UpdateAvailableEdges(int vertex, bool[] processedVertices)
+    {
+        processedVertices[vertex] = true;
+        foreach (var edge in _edges)
+        {
+            if (edge.Vertex0 == vertex && !processedVertices[edge.Vertex1] 
+                || edge.Vertex1 == vertex && !processedVertices[edge.Vertex0])
+            {
+                _availableEdges.Insert(edge);
+            }
+        }
+    }
+    
+    private void FillOutputPattern(StringBuilder[] lines)
+    {
+        for (var i = 0;  i < lines.Length; i++)
+        {
+            lines[i] = new StringBuilder();
+            lines[i].Append($"{i + 1}:");
+        }
+    }
+
+    private void Print(string targetPath, StringBuilder[] lines)
+    {
+        using (var sw = new StreamWriter(targetPath, false))
```

Comment:
```suggestion
        using var sw = new StreamWriter(targetPath, false);
```
И дальше без лишнего отступа. using со скобками имеет смысл использовать только тогда, когда ресурс надо закрыть до завершения метода.

Reply:
я так делал изначально, но после этого из 5 тестов падают 3 с ошибкой The process cannot access the file ...\output, как будто бы StreamWriter не диспозится по завершении работы метода и несколько тестов пытаются туда писать одновременно